# Minggu 1

Eksplorasi awal pengenalan dataset

Import library yang dibutuhkan<br>
salah satu library **penting** yang selalu kepakai adalah **pandas**

In [61]:
import pandas as pd
#masukkan dataset ke dalam jupyter

train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')
submission = pd.read_csv('gender_submission.csv')

Mengecek 5 data ter-atas. Bisa sesuka hati juga sih mau ngecek 7 data 10 atau berapa aja tinggal yang didalam kurung dikasih angka. Defaultnya 5

In [5]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
#untuk mengecek data terbawah
train.tail(1)

PassengerId  Survived  Pclass                 Name   Sex   Age  SibSp  \
890          891         0       3  Dooley, Mr. Patrick  male  32.0      0   

     Parch  Ticket  Fare Cabin Embarked  
890      0  370376  7.75   NaN        Q

Mengecek sebaran statistik data

In [7]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

dapat dilihat bahwa rata2 survived = 38% dari total penumpang<br>
nilai mean dan median penting untuk menjadi pertimbangan pengisian nilai NaN karena dapat mewakilkan distribusi data. Analisis lebih lanjut bisa dilakukan sendiri juga (suka-suka kita asal beralasan logis)

In [8]:
#Cek nilai yang NaN (kosong)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


Dapat dilihat bahwa total keseluruhan data adalah 891 <br>
pada kolom Age, Cabin dan Embarked memiliki nilai < 891 yg artinya terdapat nilai kosong di data tersebut.

Terdapat banyak opsi untuk mengatasi NaN ini. Pertemuan berikutnya akan dibahas

# Cari fitur yang setidaknya tidak perlu (tidak memberikan informasi yang berarti terhadap keseluruhan data)

Contohnya : PassengerId<br>
'Cabin' juga dapat di drop karena missing valuenya (NaN) lebih dari 50% data sendiri. Selain itu logikanya cabin bisa diwakilkan sama kelas gak sih kan mengelompok pasti per kelasnya.

In [62]:
train = train.drop(['PassengerId','Cabin'], axis =1)

In [10]:
#cek apakah kolom passengerId sama cabin udah ilang atau belum
train.head(1)

Survived  Pclass                     Name   Sex   Age  SibSp  Parch  \
0         0       3  Braund, Mr. Owen Harris  male  22.0      1      0   

      Ticket  Fare Embarked  
0  A/5 21171  7.25        S

Cek jumlah perbandingan wanita dan laki-laki

In [11]:
train['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

Cek perbandingan survived dan tidak survived

In [12]:
train['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Logikanya kan kalau cowok itu lebih strong jadi kemungkinan besar cowok yang banyak survived apalagi jumlah cowoknya sedikit. Kita boleh berasumsi, tapi mari kita cek

In [13]:
train.loc[train['Survived']==1].head() #loc digunakan untuk mengecek data yang lokasinya sesuai sama perintah diberi

Survived  Pclass                                               Name  \
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
8         1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9         1       2                Nasser, Mrs. Nicholas (Adele Achem)   

      Sex   Age  SibSp  Parch            Ticket     Fare Embarked  
1  female  38.0      1      0          PC 17599  71.2833        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250        S  
3  female  35.0      1      0            113803  53.1000        S  
8  female  27.0      0      2            347742  11.1333        S  
9  female  14.0      1      0            237736  30.0708        C

In [14]:
train.loc[train['Survived']==0].head()

Survived  Pclass                            Name   Sex   Age  SibSp  Parch  \
0         0       3         Braund, Mr. Owen Harris  male  22.0      1      0   
4         0       3        Allen, Mr. William Henry  male  35.0      0      0   
5         0       3                Moran, Mr. James  male   NaN      0      0   
6         0       1         McCarthy, Mr. Timothy J  male  54.0      0      0   
7         0       3  Palsson, Master. Gosta Leonard  male   2.0      3      1   

      Ticket     Fare Embarked  
0  A/5 21171   7.2500        S  
4     373450   8.0500        S  
5     330877   8.4583        Q  
6      17463  51.8625        S  
7     349909  21.0750        S

dari sample data headnya kebetulan malah menunjukkan yg lebih banyak survived adalah cewek. Tapi kan baru sample data dan cuma 5. Mari kita cek keseluruhannya

In [15]:
train.groupby('Sex').mean()

Survived    Pclass        Age     SibSp     Parch       Fare
Sex                                                                 
female  0.742038  2.159236  27.915709  0.694268  0.649682  44.479818
male    0.188908  2.389948  30.726645  0.429809  0.235702  25.523893

Huwoo ternyata benar malah ceweknya yang paling banyak survived (74%). Ladies first mungkin ya jadi karena yang laki-laki pada strong mereka rela untuk menyelamatkan para wanita terlebih dahulu. <br>
Ini baru berdarkan gender nih. Kan yang naik beragam umurnya. Dari gender keliatan kalau ternyata yang banyak survived yang cewek.<br>
Apakah umur mempengaruhi? Misal anak-anak kan harusnya juga jadi prioritas yang diselamatkan (pada prosedur penyelamatan). Mari kita cek di pertemuan berikutnya

# Mengubah Kategorikal data menjadi numerikal

Data kategorikal pada titanic ini : <br>
- Sex
- Ticket
- Name
- Embarked

In [63]:
'''ubah keterangan 'sex' ke dalam numerik.
    Female = 1
    Male = 0
'''
def gen(Gen):
    if Gen == 'male':
        return 0
    elif Gen == 'female':
        return 1

In [64]:
train['Sex'] = train['Sex'].apply(gen)

In [18]:
#cek dah keubah belum
train.head(3)

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   

    Age  SibSp  Parch            Ticket     Fare Embarked  
0  22.0      1      0         A/5 21171   7.2500        S  
1  38.0      1      0          PC 17599  71.2833        C  
2  26.0      0      0  STON/O2. 3101282   7.9250        S

# Minggu 2

In [20]:
train.head()

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1   
4         0       3                           Allen, Mr. William Henry    0   

    Age  SibSp  Parch            Ticket     Fare Embarked  
0  22.0      1      0         A/5 21171   7.2500        S  
1  38.0      1      0          PC 17599  71.2833        C  
2  26.0      0      0  STON/O2. 3101282   7.9250        S  
3  35.0      1      0            113803  53.1000        S  
4  35.0      0      0            373450   8.0500        S

- one hot encoding -> 0 0 1, 0 1 0, 1 0 0
- label encoding -> 0 1 2 3 dst

In [65]:
train = pd.get_dummies(train, columns = ['Embarked'])

In [66]:
train.head()

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1   
4         0       3                           Allen, Mr. William Henry    0   

    Age  SibSp  Parch            Ticket     Fare  Embarked_C  Embarked_Q  \
0  22.0      1      0         A/5 21171   7.2500           0           0   
1  38.0      1      0          PC 17599  71.2833           1           0   
2  26.0      0      0  STON/O2. 3101282   7.9250           0           0   
3  35.0      1      0            113803  53.1000           0           0   
4  35.0      0      0            373450   8.0500           0           0   

   Embarked_S  
0           1  
1           0  
2           1  
3           1  
4           1

In [67]:
train['Title'] = [i.split('.')[0] for i in train.Name.astype('str')]
#train['title'] = [i.split(',')[1] for i in train.title.astype('str')]
train['Title'] = [i.split(',')[1] for i in train.Title.astype('str')]


In [68]:
train.head()

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1   
4         0       3                           Allen, Mr. William Henry    0   

    Age  SibSp  Parch            Ticket     Fare  Embarked_C  Embarked_Q  \
0  22.0      1      0         A/5 21171   7.2500           0           0   
1  38.0      1      0          PC 17599  71.2833           1           0   
2  26.0      0      0  STON/O2. 3101282   7.9250           0           0   
3  35.0      1      0            113803  53.1000           0           0   
4  35.0      0      0            373450   8.0500           0           0   

   Embarked_S  Title  
0           1     Mr  
1           0    Mrs  
2           1   Miss  
3           1    Mrs  
4           1     Mr

In [42]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 14 columns):
Survived      891 non-null int64
Pclass        891 non-null int64
Name          891 non-null object
Sex           891 non-null int64
Age           714 non-null float64
SibSp         891 non-null int64
Parch         891 non-null int64
Ticket        891 non-null object
Fare          891 non-null float64
Embarked_C    891 non-null uint8
Embarked_Q    891 non-null uint8
Embarked_S    891 non-null uint8
title         891 non-null object
Title         891 non-null object
dtypes: float64(2), int64(5), object(4), uint8(3)
memory usage: 79.3+ KB


In [41]:
train.Title.unique()

array([' Mr', ' Mrs', ' Miss', ' Master', ' Don', ' Rev', ' Dr', ' Mme',
       ' Ms', ' Major', ' Lady', ' Sir', ' Mlle', ' Col', ' Capt',
       ' the Countess', ' Jonkheer'], dtype=object)

In [73]:
train['title'] = [i.split('.')[0] for i in train.Name.astype('str')]
train['title'] = [i.split(',')[1] for i in train.title.astype('str')]

In [74]:
train.head()

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1   
4         0       3                           Allen, Mr. William Henry    0   

    Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  Embarked_S  title  
0  22.0      1      0   7.2500           0           0           1     Mr  
1  38.0      1      0  71.2833           1           0           0    Mrs  
2  26.0      0      0   7.9250           0           0           1   Miss  
3  35.0      1      0  53.1000           0           0           1    Mrs  
4  35.0      0      0   8.0500           0           0           1     Mr

In [69]:
train = train.drop(['Title'], axis =1 )

In [70]:
train = train.drop(['Ticket'], axis =1 )

In [71]:
train.head(3)

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   

    Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  Embarked_S  
0  22.0      1      0   7.2500           0           0           1  
1  38.0      1      0  71.2833           1           0           0  
2  26.0      0      0   7.9250           0           0           1

In [46]:
train = train.drop(['Name'], axis =1)

In [75]:
train.head()

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1   
4         0       3                           Allen, Mr. William Henry    0   

    Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  Embarked_S  title  
0  22.0      1      0   7.2500           0           0           1     Mr  
1  38.0      1      0  71.2833           1           0           0    Mrs  
2  26.0      0      0   7.9250           0           0           1   Miss  
3  35.0      1      0  53.1000           0           0           1    Mrs  
4  35.0      0      0   8.0500           0           0           1     Mr

In [88]:
train[['title','Sex','Survived']].groupby(['title'], 
            as_index=False).mean().sort_values(by='Survived', 
                                               ascending = False)

title       Sex  Survived
0      0  0.857143  1.000000
1      1  1.000000  0.792000
2      2  1.000000  0.697802
3      3  0.000000  0.575000
4      4  0.000000  0.500000
5      5  0.000000  0.500000
6      6  0.142857  0.428571
7      7  0.000000  0.156673
8      8  0.000000  0.000000

In [76]:
def Title(t):
    if t == ' the Countess' or t == ' Mlle' or t == ' Sir' or t == ' Ms' or t ==' Lady' or t ==' Mme':
        return 0
    elif t == ' Mrs':
        return 1
    elif t == ' Miss':
        return 2
    elif t == ' Master':
        return 3
    elif t == ' Col':
        return 4
    elif t == ' Major':
        return 5
    elif t == ' Dr':
        return 6
    elif t == ' Mr':
        return 7
    else:
        return 8

train['title'] = train['title'].apply(Title)

In [78]:
train = train.drop(['Name'], axis =1)

In [79]:
train.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    0  22.0      1      0   7.2500           0           0   
1         1       1    1  38.0      1      0  71.2833           1           0   
2         1       3    1  26.0      0      0   7.9250           0           0   
3         1       1    1  35.0      1      0  53.1000           0           0   
4         0       3    0  35.0      0      0   8.0500           0           0   

   Embarked_S  title  
0           1      7  
1           0      1  
2           1      2  
3           1      1  
4           1      7

In [80]:
pd.crosstab(train['title'], train['Survived'])

Survived    0    1
title             
0           0    7
1          26   99
2          55  127
3          17   23
4           1    1
5           1    1
6           4    3
7         436   81
8           9    0

In [81]:
pd.crosstab(train['title'], train['Sex'])

Sex      0    1
title          
0        1    6
1        0  125
2        0  182
3       40    0
4        2    0
5        2    0
6        6    1
7      517    0
8        9    0

In [82]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
Survived      891 non-null int64
Pclass        891 non-null int64
Sex           891 non-null int64
Age           714 non-null float64
SibSp         891 non-null int64
Parch         891 non-null int64
Fare          891 non-null float64
Embarked_C    891 non-null uint8
Embarked_Q    891 non-null uint8
Embarked_S    891 non-null uint8
title         891 non-null int64
dtypes: float64(2), int64(6), uint8(3)
memory usage: 58.4 KB


In [85]:
def isiUmur(u):
    if u == 0:
        return train.loc[train['title']==0].Age.mean()
    elif u == 1:
        return train.loc[train['title']==1].Age.mean()
    elif u == 2:
        return train.loc[train['title']==2].Age.mean()
    elif u == 3:
        return train.loc[train['title']==3].Age.mean()
    elif u == 4:
        return train.loc[train['title']==4].Age.mean()
    elif u == 5:
        return train.loc[train['title']==5].Age.mean()
    elif u == 6:
        return train.loc[train['title']==6].Age.mean()
    elif u == 7:
        return train.loc[train['title']==7].Age.mean()
    elif u == 8:
        return train.loc[train['title']==8].Age.mean()
    
train['Age']= train.Age.fillna(train.title.apply(isiUmur))

In [94]:
round(train.loc[train['title']==0].Age.mean())

33

In [86]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
Survived      891 non-null int64
Pclass        891 non-null int64
Sex           891 non-null int64
Age           891 non-null float64
SibSp         891 non-null int64
Parch         891 non-null int64
Fare          891 non-null float64
Embarked_C    891 non-null uint8
Embarked_Q    891 non-null uint8
Embarked_S    891 non-null uint8
title         891 non-null int64
dtypes: float64(2), int64(6), uint8(3)
memory usage: 58.4 KB


In [87]:
X_train = train[:600]


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    0  22.0      1      0   7.2500           0           0   
1         1       1    1  38.0      1      0  71.2833           1           0   
2         1       3    1  26.0      0      0   7.9250           0           0   
3         1       1    1  35.0      1      0  53.1000           0           0   
4         0       3    0  35.0      0      0   8.0500           0           0   

   Embarked_S  title  
0           1      7  
1           0      1  
2           1      2  
3           1      1  
4           1      7

In [89]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(train.drop(['Survived'], 
                                                               axis =1),
                                                    train['Survived'], 
                                                    test_size=0.2)


In [91]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(712, 10) (179, 10) (712,) (179,)


In [100]:
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()
#membuat model
logreg.fit(X_train, y_train)

Prediksi = logreg.predict(X_test)
akurasi_train = round(logreg.score(X_train, y_train) *100,2)
akurasi_validasi = round(logreg.score(X_test, y_test) *100,2)

print(akurasi_train)
print(akurasi_validasi)

82.72
80.45


In [101]:
Prediksi

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [102]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [103]:
train.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    0  22.0      1      0   7.2500           0           0   
1         1       1    1  38.0      1      0  71.2833           1           0   
2         1       3    1  26.0      0      0   7.9250           0           0   
3         1       1    1  35.0      1      0  53.1000           0           0   
4         0       3    0  35.0      0      0   8.0500           0           0   

   Embarked_S  title  
0           1      7  
1           0      1  
2           1      2  
3           1      1  
4           1      7